### DISCUSSION TITLE
Code analysis: parallel parsing (and parallel type-checking)

### DISCUSSION THREAD

**Discussion by safesparrow** [dotnet/fsharp#14199] [Ideas]
# Context
## Commandline compilation performance
Parallel parsing in compilation has been around for a year and enabled by default.

Compilation can now be partially parallelised thanks to https://github.com/dotnet/fsharp/pull/13737 .

Work is also ongoing to increase parallelisation without user intervention - see:
- https://github.com/dotnet/fsharp/discussions/13733
- https://github.com/dotnet/fsharp/discussions/11634
- https://github.com/dotnet/fsharp/pull/14104
for more details.

## Code analysis performance
None of these apply to code analysis usecase (ie. the IDE).
This is a shame, since:
- users use their IDEs for analysis all the time,
- parsing and type-checking represent a bigger fraction of total wallclock time there, since lots of heavy compilation-only tasks are skipped (eg. optimisations).

## Example timeline for ParseAndCheckFileInProject in Fantomas
Using https://github.com/dotnet/fsharp/pull/13835 I produced a timeline for checking the Fantomas solution.
Full trace can be found [here](https://github.com/safesparrow/fsharp-benchmark-generator/blob/main/trace_samples/trace_fantomas.json).
The screenshot below indicates sequential nature of both parsing and type-checking, and the fact that these two are the only two significant things that happen in that call:
<img width="1919" alt="image" src="https://user-images.githubusercontent.com/2478401/198904362-b7d37a14-f31d-4ffe-b0b5-c656c47df36a.png">
This indicates that parallelisation of parsing & type-checking in the code analysis usecase can bring a huge improvement to IDE performance, much higher than it does/can for the compilation.

# Idea
The idea is simple - apply all the existing and upcoming parallelisation mechanisms to code analysis as well.

Seems like parallelising parsing should be most straightforward.
I haven't verified if there is any caching layer that avoids parsing on consecutive calls (when eg. only the first file contents changed), but even if this would only bring speedup the first time, it would be a huge benefit for many.
Personally I (have to) (re)start my IDE a lot during the day, so speeding up startup time would be great.

**Comment by T-Gro** [comment]
Yes, this is true.

With @0101 , we have the aspiration to redo the incremental build to a model more similar to Roslyn.
An immutable compilation, which is always rebuild from its previous version, and the remaining work is done on construction - in parallel where possible, never going to file system and respecting a "snapshot" nature of all its inputs.

So far, we have not spotted much opportunities to sneak into the current model of IncrementalBuild. 
More realistically, this might mean building an alternative version of IncrementalBuild from scratch (doing max. effort to keeping FsharpChecker API stable, maybe some minor additions to supply full documents instead of filenames)

**Comment by safesparrow** [comment]
Related: https://github.com/dotnet/fsharp/issues/14848 - IDE has no cache for parsing (EDIT: there is a cache, but it doesn't apply in most scenarios and is tiny by default). This is a low-hanging fruit to avoid ~10-15% of cross-file/cross-project work in FCS.

**Comment by safesparrow** [comment]
Have raised https://github.com/dotnet/fsharp/issues/14851 for parallel parsing in the service.
